hypothesis-python/src/hypothesis/internal/conjecture/shrinker.py

```diff
@@ -76,35 +80,30 @@ def defines_shrink_pass(generate_arguments):
     changes.
     """
 
-    def accept(fn):
+    run_step = attr.ib()
+    generate_arguments = attr.ib()
+
+    @property
+    def name(self):
+        return self.run_step.__name__
+
+    def __attrs_post_init__(self):
+        assert self.name not in SHRINK_PASS_DEFINITIONS, self.name
+        SHRINK_PASS_DEFINITIONS[self.name] = self
+
+
+def defines_shrink_pass(generate_arguments):
+    """A convenient decorator for defining shrink passes."""
+
+    def accept(run_step):
+        definition = ShrinkPassDefinition(
+            generate_arguments=generate_arguments, run_step=run_step
+        )
+
         def run(self):
-            # We always maintain ``arguments`` as the current collection
-            # that would be calculated from self, and instead iterate over
-            # indices into it. This means we don't have to duplicate validation
-            # logic between it and the underlying shrink pass function.
-            #
-            # Note that in the event that the shrink target actually changes
-            # this can cause us to miss a few things that we might reasonably
-            # have run. e.g. if we successfully deleted an item then every
-            # index after that point now has an off by one error. This can
-            # cause redundant work but is not a correctness issue, because
-            # we guarantee that if the collection does not change then we will
-            # iterate over every argument, and this is satisfied if the shrink
-            # target never changes when running this pass.
-            arguments = list(generate_arguments(self))
-            indices = list(hrange(len(arguments)))
-            self.random.shuffle(indices)
-            while indices:
-                i = indices.pop()
-                if i >= len(arguments):
-                    continue
-                prev = self.shrink_target
-                args = tuple(arguments[i])
-                fn(self, *args)
-                if prev is not self.shrink_target:
-                    arguments = list(generate_arguments(self))
+            return self.run_shrink_pass(definition.name)
```

Comment:
My understanding is that this (and `run_shrink_pass`) now exist only for use by tests, and are not used by the shrinker itself during normal operation.

Is this accurate? (Not a problem, just checking.)

Reply:
It is, yes.